File: go/ai/x/session.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package aix

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/firebase/genkit/go/ai"
)

// SessionState is the portable conversation state that flows between client
// and server. It contains only the data needed for conversation continuity.
type SessionState[State any] struct {
	// Messages is the conversation history (user/model exchanges).
	// Does NOT include prompt-rendered messages — those are rendered fresh each turn.
	Messages []*ai.Message `json:"messages,omitempty"`
	// Custom is the user-defined state associated with this conversation.
	Custom State `json:"custom,omitempty"`
	// Artifacts are named collections of parts produced during the conversation.
	Artifacts []*Artifact `json:"artifacts,omitempty"`
	// InputVariables is the input used for agent flows that require input variables
	// (e.g. prompt-backed agent flows).
	InputVariables any `json:"inputVariables,omitempty"`
}

// SnapshotEvent identifies what triggered a snapshot.
type SnapshotEvent string

const (
	// TurnEnd indicates the snapshot was triggered at the end of a turn.
	SnapshotEventTurnEnd SnapshotEvent = "turnEnd"
	// InvocationEnd indicates the snapshot was triggered at the end of the invocation.
	SnapshotEventInvocationEnd SnapshotEvent = "invocationEnd"
)

// SessionSnapshot is a persisted point-in-time capture of session state.
type SessionSnapshot[State any] struct {
	// SnapshotID is the unique identifier for this snapshot (UUID).
	SnapshotID string `json:"snapshotId"`
	// ParentID is the ID of the previous snapshot in this timeline.
	ParentID string `json:"parentId,omitempty"`
	// CreatedAt is when the snapshot was created.
	CreatedAt time.Time `json:"createdAt"`
	// Event is what triggered this snapshot.
	Event SnapshotEvent `json:"event"`
	// State is the actual conversation state.
	State SessionState[State] `json:"state"`
}

// SnapshotContext provides context for snapshot decision callbacks.
type SnapshotContext[State any] struct {
	// State is the current state that will be snapshotted if the callback returns true.
	State *SessionState[State]
	// PrevState is the state at the last snapshot, or nil if none exists.
	PrevState *SessionState[State]
	// TurnIndex is the turn number in the current invocation.
	TurnIndex int
	// Event is what triggered this snapshot check.
	Event SnapshotEvent
}

// SnapshotCallback decides whether to create a snapshot.
// If not provided and a store is configured, snapshots are always created.
type SnapshotCallback[State any] = func(ctx context.Context, sc *SnapshotContext[State]) bool

// SessionStore persists and retrieves snapshots.
type SessionStore[State any] interface {
	// GetSnapshot retrieves a snapshot by ID. Returns nil if not found.
	GetSnapshot(ctx context.Context, snapshotID string) (*SessionSnapshot[State], error)
	// SaveSnapshot persists a snapshot.
	SaveSnapshot(ctx context.Context, snapshot *SessionSnapshot[State]) error
}

// InMemorySessionStore provides a thread-safe in-memory snapshot store.
type InMemorySessionStore[State any] struct {
	snapshots map[string]*SessionSnapshot[State]
	mu        sync.RWMutex
}

// NewInMemorySessionStore creates a new in-memory snapshot store.
func NewInMemorySessionStore[State any]() *InMemorySessionStore[State] {
	return &InMemorySessionStore[State]{
		snapshots: make(map[string]*SessionSnapshot[State]),
	}
}

// GetSnapshot retrieves a snapshot by ID. Returns nil if not found.
func (s *InMemorySessionStore[State]) GetSnapshot(_ context.Context, snapshotID string) (*SessionSnapshot[State], error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	snap, exists := s.snapshots[snapshotID]
	if !exists {
		return nil, nil
	}

	copied, err := copySnapshot(snap)
	if err != nil {
		return nil, err
	}
	return copied, nil
}

// SaveSnapshot persists a snapshot.
func (s *InMemorySessionStore[State]) SaveSnapshot(_ context.Context, snapshot *SessionSnapshot[State]) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	copied, err := copySnapshot(snapshot)
	if err != nil {
		return err
	}
	s.snapshots[copied.SnapshotID] = copied
	return nil
}

// copySnapshot creates a deep copy of a snapshot using JSON marshaling.
func copySnapshot[State any](snap *SessionSnapshot[State]) (*SessionSnapshot[State], error) {
	if snap == nil {
		return nil, nil
	}
	bytes, err := json.Marshal(snap)
	if err != nil {
		return nil, err
	}
	var copied SessionSnapshot[State]
	if err := json.Unmarshal(bytes, &copied); err != nil {
		return nil, err
	}
	return &copied, nil
}

// SnapshotOn returns a SnapshotCallback that only allows snapshots for the
// specified events. For example, SnapshotOn[MyState](TurnEnd) will skip the
// invocation-end snapshot.
func SnapshotOn[State any](events ...SnapshotEvent) SnapshotCallback[State] {
	set := make(map[SnapshotEvent]struct{}, len(events))
	for _, e := range events {
		set[e] = struct{}{}
	}
	return func(_ context.Context, sc *SnapshotContext[State]) bool {
		_, ok := set[sc.Event]
		return ok
	}
}

// --- Session ---

// Session holds conversation state and provides thread-safe read/write access to messages,
// input variables, custom state, and artifacts.
type Session[State any] struct {
	mu    sync.RWMutex
	state SessionState[State]
	store SessionStore[State]
}

// State returns a copy of the current state.
func (s *Session[State]) State() *SessionState[State] {
	s.mu.RLock()
	defer s.mu.RUnlock()
	copied := s.copyStateLocked()
	return &copied
}

// Messages returns the current conversation history.
func (s *Session[State]) Messages() []*ai.Message {
	s.mu.RLock()
	defer s.mu.RUnlock()
	msgs := make([]*ai.Message, len(s.state.Messages))
	copy(msgs, s.state.Messages)
	return msgs
}

// AddMessages appends messages to the conversation history.
func (s *Session[State]) AddMessages(messages ...*ai.Message) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Messages = append(s.state.Messages, messages...)
}

// SetMessages replaces the entire conversation history.
func (s *Session[State]) SetMessages(messages []*ai.Message) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Messages = messages
}

// Custom returns the current user-defined custom state.
func (s *Session[State]) Custom() State {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state.Custom
}

// SetCustom updates the user-defined custom state.
func (s *Session[State]) SetCustom(custom State) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Custom = custom
}

// UpdateCustom atomically reads the current custom state, applies the given
// function, and writes the result back.
func (s *Session[State]) UpdateCustom(fn func(State) State) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Custom = fn(s.state.Custom)
}

// InputVariables returns the prompt input stored in the session state.
func (s *Session[State]) InputVariables() any {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state.InputVariables
}

// Artifacts returns the current artifacts.
func (s *Session[State]) Artifacts() []*Artifact {
	s.mu.RLock()
	defer s.mu.RUnlock()
	arts := make([]*Artifact, len(s.state.Artifacts))
	copy(arts, s.state.Artifacts)
	return arts
}

// AddArtifacts adds artifacts to the session. If an artifact with the same
// name already exists, it is replaced.
func (s *Session[State]) AddArtifacts(artifacts ...*Artifact) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, a := range artifacts {
		replaced := false
		if a.Name != "" {
			for i, existing := range s.state.Artifacts {
				if existing.Name == a.Name {
					s.state.Artifacts[i] = a
					replaced = true
					break
				}
			}
		}
		if !replaced {
			s.state.Artifacts = append(s.state.Artifacts, a)
		}
	}
}

// SetArtifacts replaces the entire artifact list.
func (s *Session[State]) SetArtifacts(artifacts []*Artifact) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state.Artifacts = artifacts
}

// copyStateLocked returns a deep copy of the state. Caller must hold mu (read or write).
func (s *Session[State]) copyStateLocked() SessionState[State] {
	bytes, err := json.Marshal(s.state)
	if err != nil {
		panic(fmt.Sprintf("agent flow: failed to marshal state: %v", err))
	}
	var copied SessionState[State]
	if err := json.Unmarshal(bytes, &copied); err != nil {
		panic(fmt.Sprintf("agent flow: failed to unmarshal state: %v", err))
	}
	return copied
}

// --- Session context ---

type sessionContextKey struct{}

type sessionHolder struct {
	session any
}

// NewSessionContext returns a new context with the session attached.
func NewSessionContext[State any](ctx context.Context, s *Session[State]) context.Context {
	return context.WithValue(ctx, sessionContextKey{}, &sessionHolder{session: s})
}

// SessionFromContext retrieves the current session from context.
// Returns nil if no session is in context or if the type doesn't match.
func SessionFromContext[State any](ctx context.Context) *Session[State] {
	holder, ok := ctx.Value(sessionContextKey{}).(*sessionHolder)
	if !ok || holder == nil {
		return nil
	}
	session, ok := holder.session.(*Session[State])
	if !ok {
		return nil
	}
	return session
}

File: py/packages/genkit/src/genkit/_ai/_agents/_session.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Agent session state and snapshot persistence."""

from __future__ import annotations

import asyncio
import weakref
from collections.abc import Awaitable, Callable
from contextvars import ContextVar
from typing import Any, Generic, Protocol, cast, runtime_checkable
from uuid import uuid4

from pydantic import BaseModel
from typing_extensions import TypeVar as TypeVarExt

from genkit._core._error import GenkitError
from genkit._core._loop_cache import _loop_local_client
from genkit._core._model import Artifact, Message, SessionSnapshot, SessionState
from genkit._core._typing import (
    SnapshotStatus,
)


def reserve_snapshot_id() -> str:
    """Mint a snapshot id that can be known before the snapshot is persisted.

    The runtime normally supplies this to the store at save time, but some flows
    need the id ahead of time — e.g. a turn that wants to name a worktree after
    the snapshot at turn start and have the snapshot at turn end reuse that id,
    or the detach path which pre-reserves the in-flight snapshot's id.
    """
    return str(uuid4())


# Custom state is a Pydantic model, so StateT is bound to BaseModel; the Any
# default covers schemaless (client-managed) sessions where custom is plain JSON.
StateT = TypeVarExt('StateT', bound=BaseModel, default=Any)
SessionContextT = TypeVarExt('SessionContextT', default=Any)
# A store only ever hands custom state back out (it's a phantom over the wire
# format), so its parameter is covariant.
StateT_co = TypeVarExt('StateT_co', covariant=True, bound=BaseModel, default=Any)


STORE_LOCK_GETTERS: weakref.WeakKeyDictionary[object, Callable[[], asyncio.Lock]] = weakref.WeakKeyDictionary()


class SessionStoreLock:
    """Loop-local asyncio.Lock for in-process stores that need one.

    Not part of :class:`SessionStore`: durable backends (e.g. Firestore) do not
    expose a public lock. In-memory and file stores mix this in for their own
    serialization.
    """

    @property
    def lock(self) -> asyncio.Lock:
        """Return a loop-local asyncio.Lock for this store instance."""
        try:
            getter = STORE_LOCK_GETTERS.get(self)
            if getter is None:
                getter = _loop_local_client(lambda: asyncio.Lock())
                STORE_LOCK_GETTERS[self] = getter
            return getter()
        except TypeError:
            # Fallback for classes that disallow weak references
            getter = getattr(self, '_loop_lock_getter', None)
            if getter is None:
                getter = _loop_local_client(lambda: asyncio.Lock())
                object.__setattr__(self, '_loop_lock_getter', getter)
            return getter()


class SessionStore(Protocol, Generic[StateT_co]):
    """Structural interface for snapshot persistence backends.

    Minimum: ``get_snapshot`` + ``save_snapshot``.
    Optional detach/abort support: implement ``SnapshotSubscriber`` as well.

    The ``StateT`` parameter names the custom-state shape a store round-trips,
    so a typed store agrees with its agent's ``state_schema``. It's a phantom
    over the snapshot wire format (which stays plain JSON), so leaving it off
    just defaults to ``Any``.
    """

    async def get_snapshot(
        self,
        *,
        snapshot_id: str | None = None,
        session_id: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> SessionSnapshot | None:
        """Retrieve a snapshot by id or the latest leaf for a session.

        ``context`` is optional request side-channel data (e.g. auth) so
        backends can isolate tenants; in-memory/file stores ignore it.
        """
        ...

    async def save_snapshot(
        self,
        snapshot_id: str,
        fn: Callable[
            [SessionSnapshot | None],
            SessionSnapshot | None,
        ],
        *,
        context: dict[str, Any] | None = None,
    ) -> SessionSnapshot | None:
        """Atomically read-modify-write a snapshot under ``snapshot_id``.

        fn receives the existing snapshot (or None for new) and returns the
        snapshot to persist, or None to skip. Stores may call fn more than
        once under contention, so it must not mutate external state; recording
        what it observed is fine as long as a retried call may overwrite the
        record (last call wins).

        Callers reserve the id up front (``reserve_snapshot_id``) so it can be
        known before the write — e.g. handed to a turn handler via
        ``TurnContext``. When no row exists yet, this creates under that id.
        The store also fills ``created_at`` and defaults status when left empty.

        ``context`` is optional request side-channel data (e.g. auth) so
        backends can isolate tenants; in-memory/file stores ignore it.
        """
        ...


@runtime_checkable
class SnapshotStatusStream(Protocol):
    """Async iterator of snapshot status changes.

    Consume with ``async for``. Call ``aclose()`` (or ``contextlib.aclosing``)
    to unsubscribe — Python does not tear subscriptions down on ``break`` alone.
    Natural end after a terminal status also unsubscribes.
    """

    def __aiter__(self) -> SnapshotStatusStream: ...

    async def __anext__(self) -> SnapshotStatus: ...

    async def aclose(self) -> None: ...


@runtime_checkable
class SnapshotSubscriber(Protocol):
    """Optional capability that makes a store's snapshots abortable/detachable.

    Aborting itself is just a ``save_snapshot`` that flips a pending snapshot to
    aborted — there's no separate abort method. This is the other half: a way to
    *notice* that flip (e.g. when a different request aborts a detached turn
    that's still running) so the runtime can cancel the background work. A store
    that can't signal status changes can't support detach.
    """

    async def on_snapshot_status_change(
        self, snapshot_id: str, *, context: dict[str, Any] | None = None
    ) -> SnapshotStatusStream:
        """Async stream of the snapshot's persisted status changes.

        Yields committed statuses such as pending, completed, failed, or aborted.
        ``expired`` is computed on read when a pending heartbeat goes stale — it
        is not written to the store and does not appear on this stream. Use
        ``get_snapshot`` / resolve paths for that liveness check.

        Iteration ends when the run resolves (the last yielded status is a
        persisted terminal) or when the store is closed locally (ends without
        one). Call ``aclose()`` (or wrap with ``contextlib.aclosing``) to
        unsubscribe early — ``break`` alone is not enough. ``context`` scopes
        the subscription for multi-tenant stores; stores without tenancy accept
        and ignore it.
        """
        ...


def select_leaf_snapshot(
    *,
    snapshots: list[SessionSnapshot],
    session_id: str,
) -> SessionSnapshot | None:
    if not snapshots:
        return None

    parent_ids = {snap.parent_id for snap in snapshots if snap.parent_id}
    leaves = [snap for snap in snapshots if snap.snapshot_id not in parent_ids]

    if len(leaves) == 1:
        return leaves[0]

    if not leaves:
        raise GenkitError(
            status='FAILED_PRECONDITION',
            message=(
                f"Session '{session_id}' has no leaf snapshot (corrupt or cyclic "
                'history). Resume by snapshot_id instead.'
            ),
        )

    raise GenkitError(
        status='FAILED_PRECONDITION',
        message=(
            f"Session '{session_id}' has branching snapshots ({len(leaves)} "
            'leaves), so there is no single latest snapshot. This happens when a '
            'conversation is branched (e.g. regenerate). Resume by '
            'snapshot_id instead.'
        ),
    )


class Session(Generic[StateT]):
    """Holds conversation state with asyncio-safe read/write access.

    Parameterize with a custom-state type when agents carry typed ``custom``
    blobs (``Session[MyState]``). Wire storage stays ``SessionState.custom``.

    ``version`` bumps on every mutation so the runtime can skip redundant
    snapshot writes without deep-comparing state.
    """

    def __init__(self, initial_state: SessionState | None = None) -> None:
        self.lock = asyncio.Lock()
        # Own a copy so minting session_id (or later mutations) never reaches
        # back into a caller's AgentInit.state / snapshot blob.
        state = initial_state.model_copy(deep=True) if initial_state is not None else SessionState()
        # Every conversation needs a stable id for trace correlation and so
        # client-managed state is self-describing from the first turn.
        if not state.session_id:
            state.session_id = str(uuid4())
        self.session_state: SessionState = state
        self.version: int = 0
        self.custom_changed_listeners: list[Callable[[], Awaitable[None]]] = []
        self.artifact_changed_listeners: list[Callable[[Artifact], Awaitable[None]]] = []

    def on_custom_changed(self, listener: Callable[[], Awaitable[None]]) -> None:
        """Register a callback invoked after ``update_custom`` mutates state."""
        self.custom_changed_listeners.append(listener)

    def on_artifact_changed(self, listener: Callable[[Artifact], Awaitable[None]]) -> None:
        """Register a callback invoked after ``add_artifacts`` mutates state."""
        self.artifact_changed_listeners.append(listener)

    async def notify_custom_changed(self) -> None:
        for listener in self.custom_changed_listeners:
            await listener()

    async def notify_artifact_changed(self, artifact: Artifact) -> None:
        for listener in self.artifact_changed_listeners:
            await listener(artifact)

    async def state(self) -> SessionState:
        """Deep copy of current state."""
        async with self.lock:
            return self.session_state.model_copy(deep=True)

    async def get_messages(self) -> list[Message]:
        async with self.lock:
            return list(self.session_state.messages or [])

    async def add_messages(self, messages: list[Message]) -> None:
        async with self.lock:
            if self.session_state.messages is None:
                self.session_state.messages = []
            self.session_state.messages.extend(messages)
            self.version += 1

    async def set_messages(self, messages: list[Message]) -> None:
        async with self.lock:
            self.session_state.messages = list(messages)
            self.version += 1

    async def get_custom(self) -> StateT | None:
        async with self.lock:
            return cast(StateT | None, self.session_state.custom)

    async def update_custom(self, fn: Callable[[StateT | None], StateT]) -> None:
        async with self.lock:
            self.session_state.custom = fn(cast(StateT | None, self.session_state.custom))
            self.version += 1
        await self.notify_custom_changed()

    async def get_artifacts(self) -> list[Artifact]:
        async with self.lock:
            return list(self.session_state.artifacts or [])

    async def add_artifacts(self, artifacts: list[Artifact]) -> None:
        """Append artifacts; replace by name if artifact.name already exists."""
        changed: list[Artifact] = []
        async with self.lock:
            if self.session_state.artifacts is None:
                self.session_state.artifacts = []
            for art in artifacts:
                replaced = False
                if art.name:
                    for i, existing in enumerate(self.session_state.artifacts):
                        if existing.name == art.name:
                            self.session_state.artifacts[i] = art
                            replaced = True
                            break
                if not replaced:
                    self.session_state.artifacts.append(art)
                changed.append(art)
            self.version += 1
        for art in changed:
            await self.notify_artifact_changed(art)


# ---------------------------------------------------------------------------
# Session context (async-local binding for middleware and tools)
# ---------------------------------------------------------------------------

current_session: ContextVar[Session[Any] | None] = ContextVar('genkit.session', default=None)


def get_current_session() -> Session[Any] | None:
    """Return the session bound by :func:`run_with_session`, if any."""
    return current_session.get()


async def run_with_session(
    *,
    session: Session[StateT],
    coro: Awaitable[SessionContextT],
) -> SessionContextT:
    """Run ``coro`` with ``session`` available via :func:`get_current_session`."""
    token = current_session.set(session)
    try:
        return await coro
    finally:
        current_session.reset(token)
